batch-setup/make_meta_tiles.py

```diff
@@ -244,10 +251,185 @@ def __call__(self, parent):
+
+    A pool size of 30 seems to work well; the point of the pool is to hide
+    the latency of S3 requests, so pretty quickly hits the law of diminishing
+    returns with extra concurrency (without the extra network cards that Batch
+    would provide).
+    """
+
+    p = Pool(pool_size)
+    job_sizer = _JobSizer(rawr_bucket, prefix, key_format_type, rawr_zoom)
 
-        return parent, size
+    grouped_by_rawr_tile_size = []
+    print("[%s] Bucketizing jobs by zoom using size of raw tiles" % time.ctime())
+    # loop over each column individually. this limits the number of concurrent
+    # tasks, which means we don't waste memory maintaining a huge queue of
+    # pending tasks. and when something goes wrong, the stacktrace isn't buried
+    # in a million others.
+    all_sizes = {}
+    grouping_queue = queue.Queue()
 
+    num_coords = 1 << group_zoom
+    for x in range(num_coords):
+        # kick off tasks async. each one knows its own coordinate, so we only
+        # need to track the handle to know when its finished.
+        tasks = []
+        for y in range(num_coords):
+            coord = Coordinate(zoom=group_zoom, column=x, row=y)
+            tasks.append(p.apply_async(job_sizer, (coord,)))
+            grouping_queue.put_nowait(coord)  # queue for future size counting
+
+        # collect tasks and put them into the big jobs list.
+        for task in tasks:
+            coord, sizes = task.get()
+            all_sizes.update(sizes)
+
+    # now use all_sizes plus the size_threshold to find the lowest zoom we can group each coordinate into
+    counts_at_zoom = {}
+    while not grouping_queue.empty():
+        this_coord = grouping_queue.get_nowait()
+        this_size = all_sizes[this_coord]
+
+        if this_size <= size_threshold or this_coord.zoom == rawr_zoom:
+            # we're good on size, or can't group at a higher zoom
+            grouped_by_rawr_tile_size.append(this_coord)
+            grouping_queue.task_done()
+
+            if this_coord.zoom not in counts_at_zoom:
+                counts_at_zoom[this_coord.zoom] = 0
+            counts_at_zoom[this_coord.zoom] += 1
+        else:
+            # too big for this zoom, queue the children
+            top_left_child = this_coord.zoomBy(1)
+
+            grouping_queue.put_nowait(top_left_child)
+            grouping_queue.put_nowait(top_left_child.down(1))
+            grouping_queue.put_nowait(top_left_child.right(1))
+            grouping_queue.put_nowait(top_left_child.down(1).right(1))
+
+    print("[%s] Done bucketizing jobs - count by zoom %s" % (time.ctime(), counts_at_zoom))
+
+    # validate counts by zoom - expecting the equivalent of 4^10 zoom 10 jobs.
+    counts_at_zoom_sum = 0
+    for z in counts_at_zoom.keys():
+        count_at_this_zoom = counts_at_zoom[z]
+        zoom_10_equiv_count = count_at_this_zoom * (4 ** (10 - z))
+        counts_at_zoom_sum += zoom_10_equiv_count
+    if counts_at_zoom_sum == 4**10:
```

Comment:
Should this `10` be one of the zoom values that gets passed in? What happens when we want to switch away from zoom 10?

Reply:
doh.  Yes definitely

Reply:
I guess it really just needs to be a number we're comfortable we're not going to go over.  It could be zoom 20 and everything would be fine (except overflow?).  At hardcoded zoom 10, if we ended up grouping into zoom 11 jobs then we'd be doing a 4^-1 calc which would then compare an int to a float, and things might get bad